RandBLAS/trig_skops.hh

```diff
@@ -0,0 +1,440 @@
+            // Apply the original FHT on this column
+            for (int64_t i = 0; i < log_n; ++i) {
+                int64_t s1 = 1 << i;
+                int64_t s2 = s1 << 1;
+                for (int64_t j = 0; j < n; j += s2) {
+                    for (int64_t k = 0; k < s1; ++k) {
+                        // For implicitly padding the input we just have to make sure
+                        // we replace all out-of-bounds accesses with zeros
+                        bool b1 = j + k < num_cols;
+                        bool b2 = j + k + s1 < num_cols;
+                        T u = b1 ? buf[(j + k) * num_rows + row] : 0;
+                        T v = b2 ? buf[(j + k + s1) * num_rows + row] : 0;
+                        if(b1 && b2) {
+                            buf[(j + k) * num_rows + row] = u + v;
+                            buf[(j + k + s1) * num_rows + row] = u - v;
+                        }
+                        else if(!b2 && b1) {
+                            buf[(j + k) * num_rows + row] = u + v;
+                        }
+                        else if(!b2 && !b1)
+                            continue;
+                    }
+                }
+            }
+        }
+    }
+
+    template <typename T>
+    void fht_dispatch(
+        bool left,
+        blas::Layout layout,
+        T* buff,
+        int64_t log_n,
+        int64_t num_rows,
+        int64_t num_cols
+        )
+    {
+        if(left && layout == blas::Layout::ColMajor)
+            fht_left_col_major(buff, log_n, num_rows, num_cols);
+        else if(left && layout == blas::Layout::RowMajor)
+            fht_left_row_major(buff, log_n, num_rows, num_cols);
+        else if(!left && layout == blas::Layout::ColMajor)
+            fht_right_col_major(buff, log_n, num_rows, num_cols);
+        else
+            fht_right_row_major(buff, log_n, num_rows, num_cols);
+    }
+}
+
+
+namespace RandBLAS::trig {
+/*
+ * These functions apply an in-place, SRHT-like transform to the input matrix
+ * i.e. A <- (\Pi H D)A OR A <- A(D H \Pi) (which is equivalent to A <- A(\Pi H D)^{-1})
+ * layout: Layout of the input matrix (`ColMajor/RowMajor`)
+ * A: (m x n), input dimensions of `A`
+ * d: The number of rows/columns that will be permuted by the action of $\Pi$
+ */
+template <typename T, typename RNG = r123::Philox4x32, SignedInteger sint_t = int64_t>
+inline void lmiget(
+    blas::Layout layout,
+    RandBLAS::RNGState<RNG> random_state,
+    int64_t m, // `A` is `(m x n)`
+    int64_t n,
+    int64_t d, // `d` is the number of rows that have to be permuted by `\Pi`
+    T* A // data-matrix
+) {
+    // Size of the Rademacher entries = |A_cols|
+    //TODO: Change `diag` to float/doubles (same data type as the matrix)
+    sint_t* diag = new sint_t[n];
+    sint_t* selected_rows = new sint_t[d];
+
+    auto [ctr, key] = random_state;
+
+    //Step 1: Scale with `D`
+        //Populating `diag`
+    generate_rademacher_vector_r123(diag, key[0], ctr[0], n);
```

Comment:
Pass in the state. Don't pass in the first component of counter and key arrays.

Reply:
Resolved!